src/client/render/core.cpp

```diff
@@ -67,6 +86,87 @@ void RenderingCore::draw(video::SColor _skycolor, bool _show_hud, bool _show_min
 	draw_crosshair = _draw_crosshair;
 
 	beforeDraw();
+
+	// todo - use a simpler shader for shadow map generation
+	static u32 shadow_shader;
+	if (shadow_shader == 0) {
+		IShaderSource *shdsrc = client->getShaderSource();
+		shadow_shader = shdsrc->getShader("shadow_shader", TILE_MATERIAL_OPAQUE, 0);
+	}
+
+	client->getEnv().getClientMap().depthTexture = NULL;
+
+	{
+		static COpenGLRenderTarget *lightScene;
+		if (lightScene == NULL) {
+#ifdef _IRR_WINDOWS_API_
+			glActiveTexture = (PFNGLACTIVETEXTUREPROC)IRR_OGL_LOAD_EXTENSION("glActiveTexture");
+#endif
+			// ARB FrameBufferObjects
+			glBindFramebuffer = (PFNGLBINDFRAMEBUFFERPROC)IRR_OGL_LOAD_EXTENSION("glBindFramebuffer");
+			glDeleteFramebuffers = (PFNGLDELETEFRAMEBUFFERSPROC)IRR_OGL_LOAD_EXTENSION("glDeleteFramebuffers");
+			glGenFramebuffers = (PFNGLGENFRAMEBUFFERSPROC)IRR_OGL_LOAD_EXTENSION("glGenFramebuffers");
+			glCheckFramebufferStatus = (PFNGLCHECKFRAMEBUFFERSTATUSPROC)IRR_OGL_LOAD_EXTENSION("glCheckFramebufferStatus");
+			glFramebufferTexture2D = (PFNGLFRAMEBUFFERTEXTURE2DPROC)IRR_OGL_LOAD_EXTENSION("glFramebufferTexture2D");
+
+			u32 SHADOWMAP_RES = 2048;
+			lightScene = new COpenGLRenderTarget(v2u32(SHADOWMAP_RES, SHADOWMAP_RES), "lightscene", true, driver);
+		}
+		//driver->setRenderTarget(0, true, true, video::SColor(0xFFFFFFFF));
+
+		// Technically, we don't need a color buffer for standard shadow mapping
+		lightScene->bindRTT(false, true);
+
+		sky.OnAnimate(porting::getTimeMs());
+		clouds.OnAnimate(porting::getTimeMs());
+		client->getCamera()->getCameraNode()->OnAnimate(porting::getTimeMs());
+		client->getEnv().getClientMap().OnAnimate(porting::getTimeMs());
+
+		//client->getEnv().getClientMap().depthTexture = NULL;
+
+		// TODO: Make sure to use sun position instead of local player position
+		v3f sunPosition = sky.getSunPosition();
```

Comment:
Biggest problem is here... we should render from sun position but this doesn't seem to work correctly. I suppose this is due to the occlusion code in `UpdateDrawList` and `RenderMapToShadowMap`.

Reply:
Sun position should be at an infinite distance, maybe you have coded it this way already but thought i'd check since MT uses a non-infinite distance for some other effects. Infinite distance might simplify the calculations because you can then calculate the shadow by using parallel light-ray lines.

Reply:
??
What do you mean exactly by "infinite distance" and "parallel light-ray lines"?
We assume that the light source is the sun and the depth map is made out from the light' point of view. The light needs to be localized.

Reply:
usually the sun is so far away the light rays are considered parallel.
https://learnopengl.com/Advanced-Lighting/Shadows/Shadow-Mapping
paragraph near the bottom of the page.
or this one here:
http://www.opengl-tutorial.org/intermediate-tutorials/tutorial-16-shadow-mapping/
near the top

Reply:
Yes, the sun or moon are directional lights so should use orthographic projection:
"directional lights - lights that are so far away that all the light rays can be considered parallel. As such, rendering the shadow map is done with an orthographic projection matrix"
Mathematically i suspect it will be much simpler so more lightweight.